presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java

```diff
@@ -0,0 +1,298 @@
+        int positionCount = 0;
+        while (true) {
+            int batchSize = orcRecordReader.nextBatch();
+            if (batchSize <= 0) {
+                break;
+            }
+            Block block = orcRecordReader.readBlock(VARCHAR, 0);
+            positionCount += block.getPositionCount();
+        }
+        assertEquals(positionCount, POSITION_COUNT);
+
+        assertEquals(orcDataSource.getCount(), 2); // read all stripes at once
+    }
+
+    public static <T, U extends T> void assertNotInstanceOf(T actual, Class<U> expectedType)
+    {
+        assertNotNull(actual, "actual is null");
+        assertNotNull(expectedType, "expectedType is null");
+        if (expectedType.isInstance(actual)) {
+            fail(String.format("expected:<%s> to not be an instance of <%s>", actual, expectedType.getName()));
+        }
+    }
+
+    static FileSinkOperator.RecordWriter createOrcRecordWriter(File outputFile, Format format, Compression compression, ObjectInspector columnObjectInspector)
+            throws IOException
+    {
+        JobConf jobConf = new JobConf();
+        jobConf.set("hive.exec.orc.write.format", format == ORC_12 ? "0.12" : "0.11");
+        jobConf.set("hive.exec.orc.default.compress", compression.name());
+        ReaderWriterProfiler.setProfilerOptions(jobConf);
+
+        Properties tableProperties = new Properties();
+        tableProperties.setProperty("columns", "test");
+        tableProperties.setProperty("columns.types", columnObjectInspector.getTypeName());
+        tableProperties.setProperty("orc.stripe.size", "1200000");
+
+        return new OrcOutputFormat().getHiveRecordWriter(
+                jobConf,
+                new Path(outputFile.toURI()),
+                Text.class,
+                compression != NONE,
+                tableProperties,
+                () -> { }
+        );
+    }
+
+    private static class FakeOrcDataSource
+            implements OrcDataSource
+    {
+        public static final FakeOrcDataSource INSTANCE = new FakeOrcDataSource();
+
+        @Override
+        public long getReadTimeNanos()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public long getSize()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void readFully(long position, byte[] buffer)
+                throws IOException
+        {
+            // do nothing
+        }
+
+        @Override
+        public void readFully(long position, byte[] buffer, int bufferOffset, int bufferLength)
+                throws IOException
+        {
+            // do nothing
+        }
+
+        @Override
+        public <K> Map<K, FixedLengthSliceInput> readFully(Map<K, DiskRange> diskRanges)
+                throws IOException
+        {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static class DelegateOrcDataSource
```

Comment:
"Delegating" is the preferred form, but let's use "Forwarding" which the term that Guava uses and also used in a couple other places in Presto.